### PR TITLE
feat: gw7833 create bookmarkfolder model and api

### PR DESCRIPTION
**Pull request by mudana-grune**
https://youtrack.weseek.co.jp/issue/GW-7833

### CODE REVIEW

File: packages/app/src/server/models/bookmark-folder.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+bookmarkFolderSchema.statics.updateBookmarkFolder = async function(bookmarkFolderId: string, name: string, parent: string):
+ Promise<BookmarkFolderDocument | null> {
+
+  const parentFolder = await this.findById(parent);
+  const updateFields = {
+    name, parent: parentFolder?._id || null,
+  };
+  const bookmarkFolder = await this.findByIdAndUpdate(bookmarkFolderId, { $set: updateFields }, { new: true });
+  return bookmarkFolder;
```
Comment (kaoritokashiki) [inline packages/app/src/server/models/bookmark-folder.ts:72]: https://mongoosejs.com/docs/tutorials/findoneandupdate.html
> You should set the new option to true to return the document after update was applied. 👏🏻

new オプション知らなかった。

File: packages/app/src/server/routes/apiv3/bookmark-folder.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      return res.apiv3Err(err, 500);
+    }
+  });
+
+  // Delete bookmark folder and children
+  router.delete('/', accessTokenParser, loginRequiredStrictly, async(req, res) => {
+    const { boookmarkFolderId } = req.body;
+    try {
+      await BookmarkFolder.deleteFolderAndChildren(boookmarkFolderId);
+      return res.apiv3();
```
Comment (kaoritokashiki) [inline packages/app/src/server/routes/apiv3/bookmark-folder.ts:71]: これは `res.apiv3();` だけを返すのはありなのだろうか

Comment (kaoritokashiki) [inline packages/app/src/server/routes/apiv3/bookmark-folder.ts:71]: `res.apiv3();` だけを返すのはベストな方法なのだろうか??

ただ、他にも同じようにしているところがある
Screen Shot 2022-10-20 at 19 25 56

**Comment by kaoritokashiki** [comment]
@mudana-grune  
Please fix lint errors

File: packages/app/src/server/models/bookmark-folder.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+const bookmarkFolderSchema = new Schema<BookmarkFolderDocument, BookmarkFolderModel>({
+  name: { type: String },
+  owner: { type: Schema.Types.ObjectId, ref: 'User', index: true },
+  parent: { type: Schema.Types.ObjectId, refPath: 'BookmarkFolder', required: false },
+});
+
+
+bookmarkFolderSchema.statics.createByParameters = async function(params: IBookmarkFolderDocument): Promise<BookmarkFolderDocument> {
+  const { name, owner, parent } = params;
+  const parentFolder = await this.findById(parent);
```
Comment (yuki-takei) [inline packages/app/src/server/models/bookmark-folder.ts:41]: This method should throw an error when the specified ID of the parent is invalid. Please create a custom error class that extends `ExtensibleCustomError`

File: packages/app/src/server/routes/apiv3/bookmark-folder.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    };
+
+    try {
+      const bookmarkFolder = BookmarkFolder.createByParameters(params);
+      logger.debug('bookmark folder created', bookmarkFolder);
+      return res.apiv3({ bookmarkFolder });
+    }
+    catch (err) {
+      logger.error('create bookmark folder failed', err);
+      return res.apiv3Err(err, 500);
```
[the remark this answers is not in the document]
Comment (yuki-takei) [reply packages/app/src/server/routes/apiv3/bookmark-folder.ts:40]: 上でコメントした createByParameters が invalid parent exception を投げる場合に対応すべき。
それ以外のエラーの場合は結局原因不明なのでこのコードでいいと思う。

File: packages/app/src/server/routes/apiv3/bookmark-folder.ts (hunk 2 above)
Comment (yuki-takei) [reply packages/app/src/server/routes/apiv3/bookmark-folder.ts:71]: delete の正常終了時はまあこれでもいいけど、欲を言えば何件削除されたとかっていう結果がわかるといいなあ。(コストが高ければ見送りで良い)

**Review by yuki-takei — CHANGES_REQUESTED** [review]

File: packages/app/src/server/models/bookmark-folder.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+  if (parent === null) {
+    bookmarkFolder = await this.create({ name, owner, parent:  null }) as unknown as BookmarkFolderDocument;
+  }
+  else {
+    const parentFolder = await this.findById(parent);
+    if (!parentFolder) {
+      throw new InvalidParentBookmarkFolder("Parent folder doesn't exists");
+    }
+    bookmarkFolder = await this.create({ name, owner, parent:  parentFolder?._id }) as unknown as BookmarkFolderDocument;
+  }
```
Comment (kaoritokashiki) [inline packages/app/src/server/models/bookmark-folder.ts:56]: I think you are misunderstanding about Yuki San's FB.

`parent` variable passed from client side might be invalid value like `111` `aaa` not mongoose object id.
That's why we need to check the value is valid and it's exactly exist.

File: packages/app/src/server/routes/apiv3/bookmark-folder.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (kaoritokashiki) [reply packages/app/src/server/routes/apiv3/bookmark-folder.ts:43]: 理解いたしました!

File: packages/app/src/server/routes/apiv3/bookmark-folder.ts (hunk 2 above)
Comment (kaoritokashiki) [reply packages/app/src/server/routes/apiv3/bookmark-folder.ts:74]: Could you return the number of deleted bookmarks?
(If it take time, you don't need to handle this.)

**Review by kaoritokashiki — CHANGES_REQUESTED** [review]

File: packages/app/src/server/models/bookmark-folder.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    bookmarkFolder = await this.create({ name, owner, parent:  null }) as unknown as BookmarkFolderDocument;
+  }
+  else {
+    // Check if parent folder id is valid and parent folder exists
+    const isParentFolderIdValid = isValidObjectId(parent);
+    const parentFolder = await this.findById(parent);
+
+    if (!isParentFolderIdValid || parentFolder == null) {
+      throw new InvalidParentBookmarkFolder("Parent folder id is invalid or parent folder doesn't exists");
+    }
```
Comment (kaoritokashiki) [inline packages/app/src/server/models/bookmark-folder.ts:58]: parentFolder should be nullable.
If you have any questions, please ask me.

File: packages/app/src/server/models/bookmark-folder.ts [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  Types, Document, Model, Schema,
+} from 'mongoose';
+
+
+import loggerFactory from '../../utils/logger';
+import { getOrCreateModel } from '../util/mongoose-utils';
+
+const logger = loggerFactory('growi:models:bookmark-folder');
+
+export class InvalidParentBookmarkFolder extends ExtensibleCustomError {}
```
Comment (kaoritokashiki) [inline packages/app/src/server/models/bookmark-folder.ts:14]: Please declare on `server/models/errors.ts` and add `Error` to the suffix.

**Review by kaoritokashiki — CHANGES_REQUESTED** [review]

File: packages/app/src/server/models/bookmark-folder.ts [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+import { isValidObjectId } from '@growi/core/src/utils/objectid-utils';
+import {
+  Types, Document, Model, Schema,
+} from 'mongoose';
+
+
+import loggerFactory from '../../utils/logger';
+import { getOrCreateModel } from '../util/mongoose-utils';
+
+import { ErrorInvalidParentBookmarkFolder } from './errors';
```
Comment (kaoritokashiki) [inline packages/app/src/server/models/bookmark-folder.ts:11]: sorry, it should be `InvalidParentBookmarkFolderError`

File: packages/app/src/server/models/bookmark-folder.ts (hunk 6 above)
Comment (kaoritokashiki) [reply packages/app/src/server/models/bookmark-folder.ts]: we can create a new folder even if the parentFolder is null. 

so when parentFolder is null, I think the error should not be thrown.

 (parentFolder != null && !isParentFolderIdValid)
i think the err should be thrown in the above case

Comment (kaoritokashiki) [reply packages/app/src/server/models/bookmark-folder.ts]: if you have any problems, please share.

**Review by kaoritokashiki — CHANGES_REQUESTED** [review]

**Review by kaoritokashiki — CHANGES_REQUESTED** [review]

**Review by kaoritokashiki — APPROVED** [review]

**Pull request merged**